File: android/app/src/main/kotlin/dev/alpaka/projectathens/admob/SpeechNativeAdFactory.kt
package dev.alpaka.projectathens.admob

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import dev.alpaka.projectathens.R
import dev.alpaka.projectathens.databinding.AdSpeechBinding
import dev.alpaka.projectathens.ext.isInvisible
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

class SpeechNativeAdFactory(private val context: Context) : GoogleMobileAdsPlugin.NativeAdFactory {
    override fun createNativeAd(nativeAd: NativeAd, customOptions: MutableMap<String, Any>?): NativeAdView = AdSpeechBinding.inflate(LayoutInflater.from(context)).let { binding ->
        binding.smallAdText.isInvisible = nativeAd.icon == null
        binding.largeAdText.isInvisible = nativeAd.icon != null
        nativeAd.icon?.let { binding.adIcon.setImageDrawable(it.drawable) }

        binding.adHeadline.text = nativeAd.headline
        binding.adBody.text = nativeAd.body

        binding.root.apply {
            iconView = binding.adIcon
            headlineView = binding.adHeadline
            bodyView = binding.adBody
            setNativeAd(nativeAd)
        }
    }
}

File: ios/Runner/SpeechNativeAdFactory.swift

import google_mobile_ads

class SpeechNativeAdFactory : FLTNativeAdFactory {

    func createNativeAd(_ nativeAd: GADNativeAd,
                        customOptions: [AnyHashable : Any]? = nil) -> GADNativeAdView? {
        let nibView = Bundle.main.loadNibNamed("SpeechNativeAd", owner: nil, options: nil)!.first
        let nativeAdView = nibView as! GADNativeAdView
        

        (nativeAdView.headlineView as! UILabel).font = UIFont.boldSystemFont(ofSize: 18)
        (nativeAdView.headlineView as! UILabel).text = nativeAd.headline

        (nativeAdView.bodyView as! UILabel).text = nativeAd.body
        nativeAdView.bodyView!.isHidden = nativeAd.body == nil

        (nativeAdView.iconView as! UIImageView).image = nativeAd.icon?.image
        nativeAdView.iconView!.isHidden = nativeAd.icon == nil

        nativeAdView.callToActionView?.isUserInteractionEnabled = false

        nativeAdView.nativeAd = nativeAd

        return nativeAdView
    }
}
